OCPP 2.0.1 support

ars3niy opened the pull request:
I made this a while ago so there are lots of merge conflicts now, but I'd like to get feedback before updating anything.

This adds OCPP 2.0.1 support (you choose protocol version when creating the integration). It involves major refactoring so I'm not sure if I broke OCPP 1.6, don't really know how to test. Only tested with an OCPP 2.0.1 charger.

drc38 commented:
Including OCPP 2.0.1 will be a great addition! If you resolve the merge conflicts the tests will run and confirm whether anything in the OCPP 1.6 implementation has been broken.

Then the next step would be to include similar tests for OCPP 2.0.1

ars3niy commented:
I'll start updating it then. I think it requires feature freeze, though, because of extensive refactoring (commits like this one https://github.com/lbbrhzn/ocpp/pull/1381/commits/c65141295cba7b96cb50a1f84c3b9dfcd8ae99ff)

drc38 commented:
> I'll start updating it then. I think it requires feature freeze, though, because of extensive refactoring (commits like this one [c651412](https://github.com/lbbrhzn/ocpp/commit/c65141295cba7b96cb50a1f84c3b9dfcd8ae99ff))

There doesn't appear to be anyone working on getting the multiple connector code ready to integrate so the only reason I can think of for the integration to be updated would be if HA or a dependency introduces a breaking issue. But this would apply to your fork also.

lightheaded commented:
> > I'll start updating it then. I think it requires feature freeze, though, because of extensive refactoring (commits like this one [c651412](https://github.com/lbbrhzn/ocpp/commit/c65141295cba7b96cb50a1f84c3b9dfcd8ae99ff))
> 
> There doesn't appear to be anyone working on getting the multiple connector code ready to integrate so the only reason I can think of for the integration to be updated would be if HA or a dependency introduces a breaking issue. But this would apply to your fork also.

Hey! I just forked the PR for multiple chargers support yesterday. Hoping to contribute. However, my pace will be quite slow. Not sure if this changes anything, but ought you should know.

drc38 commented:
> > > I'll start updating it then. I think it requires feature freeze, though, because of extensive refactoring (commits like this one [c651412](https://github.com/lbbrhzn/ocpp/commit/c65141295cba7b96cb50a1f84c3b9dfcd8ae99ff))
> > 
> > 
> > There doesn't appear to be anyone working on getting the multiple connector code ready to integrate so the only reason I can think of for the integration to be updated would be if HA or a dependency introduces a breaking issue. But this would apply to your fork also.
> 
> Hey! I just forked the PR for multiple chargers support yesterday. Hoping to contribute. However, my pace will be quite slow. Not sure if this changes anything, but I thought you should know.

@lightheaded I suggest holding off as the 2.0.1 changes will create significant rework for you otherwise.

lightheaded commented:
Makes sense, thanks for the heads up. Any ETA for this upgrade?

ars3niy commented:
I've rebased onto latest main and will now start working on test coverage.

How long this will take depends on who's slower, me to make further changes or project maintainer to review. I am, however, getting paid to do this by DEFA (a manufacturer of, among other things, an OCPP 2.0.1 charger which is also marketed for private use) so I will do my part in timely manner.

lbbrhzn commented:
thanks @ars3niy! I get paid to do other things than this project, so my track record on reviewing is not that good. But i'll try to keep up with you!

drc38 commented:
> Oh, the CI checks don't run automatically. I'll try to figure out how to run the locally then.

Using the vscode devcontainer is the best way to go.

ars3niy commented:
This new revision is supposed to pass all checks. Test coverage for OCPP 2.0.1 has not yet been added.

drc38 commented:
> This new revision is supposed to pass all checks. Test coverage for OCPP 2.0.1 has not yet been added.

Nice work, looks like the 1.6 implementation is still functional and passing the tests. I'll have a closer look at the changes over the weekend.

lightheaded commented:
Hey! Great work on this! Do you have a guesstimate on when this might be merged?

drc38 commented:
> @drc38 any progress on review?
> 
> I could do a merge and do a prerelease. The test coverage is not great though...

Hi @lbbrhzn, see above for my review. @ars3niy is yet to respond. Tests for 2.0.1 are missing which is the main barrier to it being merged.

lbbrhzn commented:
> > @drc38 any progress on review?
> > I could do a merge and do a prerelease. The test coverage is not great though...
> 
> Hi @lbbrhzn, see above for my review. @ars3niy is yet to respond. Tests for 2.0.1 are missing which is the main barrier to it being merged in my view.

@drc38 Maybe I could do a pre-release on this branch? It would allow some people to test OCPP 2.0.1 in the field...

drc38 commented:
> > > @drc38 any progress on review?
> > > I could do a merge and do a prerelease. The test coverage is not great though...
> > 
> > 
> > Hi @lbbrhzn, see above for my review. @ars3niy is yet to respond. Tests for 2.0.1 are missing which is the main barrier to it being merged in my view.
> 
> @drc38 Maybe I could do a pre-release on this branch? It would allow some people to test OCPP 2.0.1 in the field...

Up to you, it would then be easier to continue with changes to the main branch. If you do also update manifest so that websockets <14 do to a breaking change in latest release.

ars3niy commented on custom_components/ocpp/ocppv16.py:
```diff
@@ -0,0 +1,973 @@
+            self._requires_reboot = True
+            await self.notify_ha(f"A reboot is required to apply {key}={value}")
+
+    async def async_update_device_info_v16(self, boot_info: dict):
```
It's a wrapper for async_update_device_info from base ChargePoint class. The reason this and async_update_device_info_v201 exist is because they each take BootNotification JSON object of respective OCPP version as an argument and those JSON objects have slightly different format.

drc38 replied:
@lbbrhzn, what's your view? My logic is because these are not specific to the ocpp protocol they should be in the base class. The change in dict names between versions can be handled with an if statement or similar.

ars3niy commented on custom_components/ocpp/ocppv16.py:
```diff
@@ -0,0 +1,973 @@
+            boot_info.get(om.firmware_version.name, None),
+        )
+
+    def process_phases(self, data):
```
Perhaps more broadly, there could be some unification between handling OCPP 1.6 meter values and OCPP 2.0 transaction events. I will look into this but since the logic is a bit complicated, it will take some time.

ars3niy commented on custom_components/ocpp/switch.py:
```diff
@@ -56,7 +56,15 @@ class OcppSwitchDescription(SwitchEntityDescription):
         off_action=HAChargerServices.service_availability.name,
         metric_state=HAChargerStatuses.status_connector.value,
-        metric_condition=[ChargePointStatus.available.value],
+        metric_condition=[
```
> It also enables the switch to be turned on in these other states.

Whereby nothing happens, I presume, because the charger was already enabled and will still have the vehicle connected?

I'll just revert this change. It's mostly there because this whole thing started as an internal demo before being promoted to upstream contribution. I don't actually feel any need to change this behaviour right now.

ars3niy commented on custom_components/ocpp/translations/i-default.json:
```diff
@@ -60,5 +61,25 @@
+        "profile_not_supported": {
+            "message": "Setting custom charging profiles not supported"
+        },
+        "dual_profile_not_supported": {
```
I think I can get rid of both profile_not_supported and this. The former can probably just be implemented by converting OCPP 1.6 profile to 2.0 profile, and the latter cannot even happen because nothing calls set_charge_rate with both limit_amps and limit_watts set.

ars3niy commented:
Changes in new revision:
- Do not offer to choose OCPP version when creating integration. Instead, tell the charger both 1.6 and 2.0.1 is supported unless specified otherwise by setting CONF_SUBPROTOCOL in some config file.
- Do not change how charger availability is interpreted.
- Do not add new error messages for setting charging profile, and implement custom charging profiles by passing the verbatim to OCPP call just like OCPP 1.6 implementation does.

I will now start adding test coverage for the new code.

ars3niy commented:
I am unable to write tests for OCPP 2.0.1 because of a web socket glitch I do not understand. It may be the same thing I see with a real charger: shortly after the charger connects to HA, web socket drops, then the charger reconnects and works normally from there.

Any clues as to why my test fails?

drc38 commented:
> I am unable to write tests for OCPP 2.0.1 because of a web socket glitch I do not understand. It may be the same thing I see with a real charger: shortly after the charger connects to HA, web socket drops, then the charger reconnects and works normally from there.
> 
> Any clues as to why my test fails?

Have you pinned the websockets version to <14? Code changes are required to get 14.1 to work.

drc38 commented:
> I am unable to write tests for OCPP 2.0.1 because of a web socket glitch I do not understand. It may be the same thing I see with a real charger: shortly after the charger connects to HA, web socket drops, then the charger reconnects and works normally from there.
> 
> Any clues as to why my test fails?

A couple of things you can try are enabling debug for websockets to see if there's anymore detail on why/when the websockets close is occurring. Reduce the number of OCPP messages being sent until you can isolate the one causing the issue. The issue appears to be in the post_connect phase.

ars3niy commented:
> Have you pinned the websockets version to <14?

I did not do anything but I happen to have version 12.0

> you can try are enabling debug for websockets

how does one do that?

drc38 commented:
https://docs.pytest.org/en/stable/how-to/logging.html

drc38 commented:
Are you running tests in parallel on the same websocket?  If so then they can interfere with each eg close the socket.

ars3niy commented:
> Are you running tests in parallel on the same websocket? If so then they can interfere with each eg close the socket.

Don't think so, I'm only running the 2.0.1 test case when I'm testing locally:

pytest --log-level=DEBUG --cov=./ --cov-report=term-missing --timeout=30 --durations=10 -n auto -p no:sugar -rA tests/test_charge_point_v201.py::test_cms_responses_v201

drc38 commented:
Can't spot anything obvious, suggest using breakpoints to try and narrow the issue down further. Also as v1.6 does not have the same problem look for any differences in approach to the tests.

ars3niy commented:
I was able to get complete call trace. It happens when post_connect calls async_update_entry here https://github.com/ars3niy/ha-ocpp/blob/ocpp20/custom_components/ocpp/chargepoint.py#L320 (exact same logic is present on the current main branch).

It is possible this also happens with OCPP 1.6, only existing tests do not see it because they do not wait for post_connect to complete. I, however, do not want to not wait for post_connect to complete in my tests.

drc38 commented:
> I was able to get complete call trace. It happens when post_connect calls async_update_entry here https://github.com/ars3niy/ha-ocpp/blob/ocpp20/custom_components/ocpp/chargepoint.py#L320 (exact same logic is present on the current main branch).
> 
> It is possible this also happens with OCPP 1.6, only existing tests do not see it because they do not wait for post_connect to complete. I, however, do not want to not wait for post_connect to complete in my tests.

So do the tests pass if you comment out that line?

ars3niy commented:
> So do the tests pass if you comment out that line?

Yes, including when running all the tests and not just the now v2.0.1 one.

ars3niy commented:
I am able to avoid this in the test case if I set CONF_MONITORED_VARIABLES in the initial MockConfigEntry to exactly the list of measurands returned by the charger. This is not a realistic situation but it gets the job done for the time being.

ars3niy commented:
One more change addressing the last major review comment. Now meter values from both OCPP versions are processed using one and the same function.

In doing so I have sacrificed rounding logic so meter values may look ugly (0.30000000000004 etc.). I'll check that later but other than this, quality should be good.

ars3niy commented:
> > I am able to avoid this in the test case if I set CONF_MONITORED_VARIABLES in the initial MockConfigEntry to exactly the list of measurands returned by the charger. This is not a realistic situation but it gets the job done for the time being.
> 
> Based on your observations I'd also shift the entry update lines of code to after `self.post_connect_success = True` which avoids a bit of repetition after the reload.

Please do :-)

drc38 commented:
@ars3niy great work! Can you fix the couple of test deprecation warnings please.

drc38 commented:
> I am able to avoid this in the test case if I set CONF_MONITORED_VARIABLES in the initial MockConfigEntry to exactly the list of measurands returned by the charger. This is not a realistic situation but it gets the job done for the time being.

Another approach is to catch the `websockets.exceptions.ConnectionClosedOK` the update entry raises and pass or supress it.

drc38 reviewed (approved):
@lbbrhzn LGTM

drc38 commented:
@ars3niy is there anything else or ok to merge?

ars3niy commented:
I just need to test with a real charger to make sure all decimal values are OK.

ars3niy commented:
There, now fully tested by me. Rounding was needed in one place, I added it. Also changed datetime.now() call, renamed charge points in tests and refactored the new tests a little.

Test coverage check fails, however, because the commits refactor most of existing code that did not quite have full coverage to start with.
